pkg/apis/resource/common.go

```diff
@@ -55,44 +49,83 @@ func DeleteResources(req CollectionDeleteRequest, mc model.ClientSet, kubeConfig
 			Deployer: dp,
 		}
 
-		for _, s := range resources {
-			if err := pkgresource.SetSubjectID(req.Context, s); err != nil {
-				return err
-			}
-
-			s, err = tx.Resources().UpdateOne(s).
-				SetAnnotations(s.Annotations).
-				Save(req.Context)
+		for _, entities := range environmentIDToResources {
+			entities, err = pkgresource.ReverseTopologicalSortResources(entities)
 			if err != nil {
 				return err
 			}
 
-			err = pkgresource.Destroy(req.Context, tx, s, destroyOpts)
-			if err != nil {
-				return err
+			for _, s := range entities {
+				if err := pkgresource.SetSubjectID(req.Context, s); err != nil {
+					return err
+				}
+
+				s, err = tx.Resources().UpdateOne(s).
+					SetAnnotations(s.Annotations).
+					Save(req.Context)
+				if err != nil {
+					return err
+				}
+
+				err = pkgresource.Destroy(req.Context, tx, s, destroyOpts)
+				if err != nil {
+					return err
+				}
 			}
 		}
 
 		return nil
 	})
 }
 
-func UpgradeResources(req CollectionRouteUpgradeRequest, mc model.ClientSet, kubeConfig *rest.Config) error {
-	// Make sure the resources are upgraded in topological order.
-	entities, err := pkgresource.TopologicalSortResources(req.Model())
+// getResourcesGroupByEnvironment queries and groups resources by environment ID.
+func getResourcesGroupByEnvironment(
+	tx *model.Tx,
+	ids []object.ID,
+	ctx context.Context,
+) (map[object.ID][]*model.Resource, error) {
+	resources, err := tx.Resources().Query().
+		Where(resource.IDIn(ids...)).
+		All(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	for _, entity := range entities {
-		entity.ChangeComment = req.ChangeComment
+	environmentIDToResources := make(map[object.ID][]*model.Resource)
+	for _, r := range resources {
+		environmentIDToResources[r.EnvironmentID] = append(environmentIDToResources[r.EnvironmentID], r)
+	}
+
+	return environmentIDToResources, nil
+}
 
-		if err = upgrade(req.Context, kubeConfig, mc, entity, req.Draft); err != nil {
+func UpgradeResources(req CollectionRouteUpgradeRequest, mc model.ClientSet, kubeConfig *rest.Config) error {
+	err := mc.WithTx(req.Context, func(tx *model.Tx) error {
+		environmentIDToResources, err := getResourcesGroupByEnvironment(tx, req.IDs(), req.Context)
```

Comment:
The resources will not receive request resources update data `req.Model()` if we query from DB.